text/0000-auto-unlock-with-safes.md

```diff
@@ -0,0 +1,133 @@
+- Proposal Name: auto-unlock-with-safes
+- Start Date: 2019-12-10
+- NEP PR: [nearprotocol/neps#26](https://github.com/nearprotocol/neps/pull/26)
+- Issue(s):  [nearprotocol/neps#23](https://github.com/nearprotocol/neps/pull/23), [nearprotocol/neps#24](https://github.com/nearprotocol/neps/pull/24)
+
+# Summary
+[summary]: #summary
+
+Introducing a new concept of safes that allows to securely lock some data from a contract with
+automatic unlock mechanism.
+
+# Motivation
+[motivation]: #motivation
+
+There are a few NEPs that proposed solutions to address the cross-shard communication problem.
+For example when an decentralized exchange tries to swap 2 tokens without owning them.
+
+One solution is to introduce locks with automatic unlock capabilities, but without explicitly exposing the locked structure.
+While it solves the issue for a simple exchange use-case. It has some limitations and complexity for non trivial use cases.
+
+It might lead to an unexpected behavior when a token is behind the proxy contract.
+
+This proposal is to introduce explicit locked data storage which we call a `safe` that can't be copied and always resolved at the end.
+It's a familiar concept for an asynchronous development similar to guards.
+When a guard is released, the destructor (or Drop in Rust) is called and the lock can be resolved.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We introduce a new concept which we call a `safe`.
+
+- Only the owner contract can create a safe.
+- The safe has one owner.
+- A safe always has a promise attached on it that calls the owner to release the safe (unlock). 
+- Safe can't be cloned or copied, so there are always only 1 instance of a particular safe.
+- Non-owner contracts can pass safes to other contracts and read the content, but can't modify the content of the safe.
+- Owner contract can modify the content of the safe, but not the automatic release promise.
+
+### Example:
+
+When a decentralized exchange tries to move some tokens, it first has to acquire and lock the funds.
+- Dex calls token contract to lock 1000 tokens for `alice`.
+- Token contract creates a safe with the following data:
+```
+{
+  owner_id: "alice",
+  amount: 1000,
+  locked_by: "dex",
+}
+```
+- Token contract creates a promise to release the safe, e.g. by calling `unlock` on itself.
+- Token returns OK and this safe to the `dex`.
+
+Now Dex has this safe from the token contract.
+Dex can read the content of the safe and assert the content is correct.
+
+- Dex calls `transfer` on the token contract and pass this safe with this promise.
+- Token contract reads the content of the safe that it received and transfers the required amount to the new owner.
+- Token contract modifies the content of the safe and decreased the safe amount by the transferred amount. E.g.
+- Token contract can now return OK and the safe back to `dex`.
+```
+{
+  owner_id: "alice",
+  amount: 100,
```

Comment:
shouldn't it be `1000`?

Reply:
I though the transfer was for 900. Modified